File: Scripts/2021/euler_rotation.m
    function R = euler_rotation(sequence, angles)
% R = euler_rotation(sequence, angles) takes as inputs:
%   -sequence: a string which specifies the axes along which rotation
%              occurs
%   -angles: The radiants (or symbolics) of the three rotations
% and outputs:
%   -R: The desired rotation
% Euler rotations work about moving-axes
    
    if strlength(sequence) ~= 3
        disp("Sequence not valid, must be of lenght three.")
        return;
    end
    
    sequence = lower(char(sequence));
    if (sequence(2) == sequence(1) || sequence(2) == sequence(3))
        disp("Two consecutive rotation along the same axis are not valid.")
        return
    end
    
    R = elem_rot_mat(sequence(1), angles(1)) * elem_rot_mat(sequence(2), angles(2)) * elem_rot_mat(sequence(3), angles(3));

end

File: Scripts/2025/python_scripts/robobox/euler_rotation.py
import numpy as np

def euler_rotation (sequence: str, angles: list) -> np.ndarray:
    """
    Computes the rotation matrix of Euler angles

    @param sequence: The rotation sequence (e.g., 'xyz', 'yxy')
    @param angles: The list of rotations angles
    """
    # Assertions
    if len(sequence) != 3:
        raise ValueError("Sequence is not valid, must be of size 3.")
    
    sequence = sequence.lower()
    if sequence[1] == sequence[0] or sequence[1] == sequence[2]:
        raise ValueError("Two consecutive rotations along the same axis is incorrect.")
    
    def elementary_rotation(axis: str, theta: float) -> np.ndarray:
        """
        Computes the elementary rotation axis along a given axis

        @param axis: The string of the provided axis (e.g., 'x', 'y', 'z')
        @param: The angle value
        """
        c, s = np.cos(theta), np.sin(theta)
        if axis == 'x':
            return np.array([[1, 0, 0],
                             [0, c, -s],
                             [0, s, c]])
        elif axis == 'y':
            return np.array([[c, 0, s],
                             [0, 1, 0],
                             [-s, 0, c]])
        elif axis == 'z':
            return np.array([[c, -s, 0],
                             [s, c, 0],
                             [0, 0, 1]])
        else:
            raise ValueError(f"Invalid axis: {axis}. Must be 'x', 'y', or 'z'.")
    
    # Computing of the rotation matrix
    R = np.eye(3)
    for axis, angle in zip(sequence, angles):
        R = R @ elementary_rotation(axis, angle)

    return R